igvc_navigation/include/igvc_navigation/PriorityQueue.h

```diff
@@ -0,0 +1,213 @@
+/**
+Priority Queue implementation for Field D* path planning algorithm.
```

Comment:
Yeah. It implements a priority queue to deal with the lexicographic order required by D* Lite and Field D*. It also implements a custom comparative to enforce  relative ordering. I'll put this in the header.

Reply:
I may be missing something, but std::priority_queue lets you define your own comparator to order your priority queue however you please. Why does your custom sorting not fit into that mechanism?

Reply:
If I'm not mistaken the priority queue here is really just a sorted set with the methods renamed to match the priority queue ones, and not really a priority queue since it implements an arbitrary remove and contains

Reply:
I created this priority queue to implement the specific methods required by the incremental path planners. Some of these methods are those that would not be available in a regular `std::priority_queue`, i.e. `Node PriorityQueue::topNode()`, `PriorityQueue::topKey()`. Additionally, since the priority queue is specialized to handle nodes and their lexicographically-ordered keys, I defined the `Key` struct and corresponding Comparator to be used by the priority queue. In short, I find my specialization of `std::priority_queue` to make sense in the context of the graph search problem and as such, easier to interpret.

Reply:
After looking through the implementation of this yesterday, I agree with @oswinso. The things that make your container different (and justify its existence) are:
- the remove method
- the contains method

topKey and topNode are simply convenience methods over the standard library (which would look more like pq.top().key() or pq.top().node().

I want to stress that, while it is certainly useful for the graph search problem, the custom method of sorting elements is not something that justifies the creation of a new container. If this was the only thing different, you would be better off using standard components.

Reply:
I agree that the features my implementation adds over the standard priority_queue are light. But, these components are used extensively throughout the path planning process. Knowing this, I think using this container abstracts away functionality that would look out of place in, say, `DLitePlanner.cpp`. i.e.: 

```C++
PQ.remove(Node(2,2));
```

instead of:
```C++
std::set<std::pair<Node, Key>>::iterator it = std::find_if(
      pq.begin(), pq.end(), [n](std::pair<Node, Key> const& e) { return e.first.getIndex() == n.getIndex(); });

if (it == pq.end())  // no such item exists
   return false;

pq.erase(it);
```

Reply:
Yes, I am agreeing with you on this part. These are features that are not only more succinct in your implementation, but are straight up not possible in the standard container (std::priority_queue does not expose iterators, so there is no find or erase mechanism).

Reply:
Why are you using `std::find_if` and a lambda? 
Is this not simply
```c++
auto it = std::find(Node(2, 2));
if (it == pq.end()) {
  return false;
}
pq.erase(it);
```

Reply:
No, because he’s comparing pair objects using only the index member of the Node.

Reply:
oh, that's true. mb